Deploying_Seq2Seq_Model_with_Hybrid_Frontend.ipynb

```diff
@@ -0,0 +1,720 @@
+    "\n",
+    "### Hybrid Frontend Notes:\n",
+    "\n",
+    "The `forward` method of this module involves iterating over the range of $[0, max\\_length)$ when decoding an output sequence one word at a time. Because of this, we should use **scripting** to convert this module to Torch Script. Unlike with our encoder and decoder models, which we can trace, we must make some necessary changes to the `GreedySearchDecoder` module in order to initialize an object without error. In other words, we must ensure that our module adheres to the rules of the scripting mechanism, and does not utilize any language features outside of the subset of Python that Torch Script includes.\n",
+    "\n",
+    "To get an idea of some manipulations that may be required, we will go over the diffs between the `GreedySearchDecoder` implementation from the chatbot tutorial and the implementation that we use in the cell below. Note that the lines highlighted in red are lines removed from the original implementation and the lines highlighted in green are new.\n",
+    "\n",
+    "![diff](images/gsd_diff.png)\n",
+    "\n",
+    "\n",
+    "#### Changes:\n",
+    "\n",
+    "- `nn.Module` -> `torch.jit.ScriptModule`\n",
+    "\n",
+    "    - In order to use PyTorch's scripting mechanism on a module, that module must inherit from the `torch.jit.ScriptModule`.\n",
+    "\n",
+    "\n",
+    "- Added `decoder_n_layers` to the constructor arguments\n",
+    "\n",
+    "    - This change stems from the fact that the encoder and decoder models that we pass to this module will be a child of `TracedModule` (not `Module`). Therefore, we cannot access the decoder's number of layers with `decoder.n_layers`. Instead, we plan for this, and pass this value in during module construction.\n",
+    "\n",
+    "\n",
+    "- Store away new attributes as constants\n",
+    "\n",
+    "    - In the original implementation, we were free to use variables from the surrounding (global) scope in our `GreedySearchDecoder`'s `forward` method. However, now that we are using scripting, we do not have this freedom. An easy solution to this is to store these values from the global scope as attributes to the module in the constructor, and add them to a special list called `__constants__` so that they can be used as literal values when constructing the graph in the `forward` method. An example of this usage is on NEW line 19, where instead of using the `device` and `SOS_token` global values, we use our constant attributes `self._device` and `self._SOS_token`.\n",
```

Comment:
Mention that we cannot do this because the assumption with scripting is that we cannot necessarily hold on to the Python objects, especially if it goes into export (which hasn't been introduced, but oh well)